### ISSUE TITLE
Support non Int64 primary keys

### ISSUE THREAD

**Issue by jhass** [amberframework/granite-orm#92]
There are other valid primary key types, such as UUIDs (mapped to `String`)

**Comment by drujensen** [comment]
@jhass What about adding a secondary unique index for the UUID instead of replacing the existing id?  This is how I usually deal with UUID's or slugs.  Or is this for backward compatibility?

Also, is this something you are interested in working on?

**Comment by jhass** [comment]
Why should I make it secondary if I want to keep it unique and use as only external and internal identifier? If I don't want predictable IDs to leak the best is to just not have them in the first place.

Unfortunately I'm not sure I can promise to work on this.

**Comment by drujensen** [comment]
Not sure what you mean, you can make multiple unique identifiers.  What is wrong with maintaining ID and UUID?

You can keep the ID as Int64 and add a UUID that is a String and add a before filter to generate a new one on create.  You can add a unique index on the UUID so searching will be fast.  You can perform `find_by :uuid, "some-generated-uuid"` in your controllers instead of using the `find id`.

Let me know why replacing the ID instead of adding a secondary unique id would be needed.  We can look into providing this functionality but I think there is no compelling reason to add that complexity when a secondary key is perfectly reasonable solution imo.

**Comment by jhass** [comment]
I disagree that it's reasonable to maintain two identifiers for the same purpose simply because the ORM lacks functionality. 

Having to maintain two identifiers adds enourmous technical debt to the project. I know that I don't want to expose or even really want to use one of them, but does every future developer? Also don't disregard the added uncertainty of what to use in what situation, especially around associations. 

Besides there's of course the argument of mapping existing database schemas, if it helps anything read only access to a database is also far from unseen. This is both not my usecase, so I won't argue that point to death, but I'm convinced it's entirely valid.

**Comment by drujensen** [comment]
ok, thanks for your input.  I can see the duplicate id as being unnecessary overhead.  The complexity is high to allow for different id generators and I'm not completely convinced a secondary is not a good alternative but I can definitely see backward compatibility as being necessary.

If you are interested in helping on this issue, your contributions are welcome.

**Comment by faustinoaq** [comment]
Maybe we can try to implement something like:

```crystal
class Users < Granite::ORM
  default_id false

  field id : String
  # ...
end
```

**Comment by drujensen** [comment]
This has been addressed by #166
